include/pybind11/pytypes.h

```diff
@@ -613,6 +615,30 @@ struct tuple_item {
         }
     }
 };
+
+struct dict_item {
+    using key_type = object;
+
+    static object get(handle obj, handle key) {
+        if (PyObject *result = PYBIND11_DICT_GET_ITEM_WITH_ERROR(obj.ptr(), key.ptr())) {
+            return reinterpret_borrow<object>(result);
+        } else {
+            // NULL with an exception means exception occurred when calling
+            // "__hash__" or "__eq__" on the key
+            // NULL without an exception means the key wasn’t present
+            if (!PyErr_Occurred())
+                // Synthesize a KeyError with the key
+                PyErr_SetObject(PyExc_KeyError, key.ptr());
```

Comment:
Is this the right error to set? Should we talk about `__hash__` or `__eq__`?

Should we even care about throwing from these two? In C++, a throw from `std::hash<T>` is just plain insane. What's the state of `std::unordered_map` after a throw from `std::hash` or `std::equal`?

Reply:
As noted, there's two kinds of errors:
1) `NULL` returned but not `PyErr_Occurred()`: key is not present (we have to set an error ourselves)
2) `NULL` returned and `PyErr_Occurred()`: exception occurred during `__hash__` or `__eq__` (and the error is already set).

Reply:
Right. My bad. Though the second part is important too.

> Should we even care about throwing from these two? In C++, a throw from std::hash<T> is just plain insane. What's the state of std::unordered_map after a throw from std::hash or std::equal?

The problem is not in inserting a single element and throwing. The problem is rehashing the dictionary and throwing half-way during the operation. At that point, basically anything goes. Thus, should we care about

> 1. `NULL` returned but not `PyErr_Occurred()`: key is not present

Reply:
Rgiht, I kind of missed that. But anyway, yes, Python throws an exception so we need to propagate this as C++ exception. We're not going to change how Python deals with this? I think the current implementation follows accessing `dict`s in pure Python.